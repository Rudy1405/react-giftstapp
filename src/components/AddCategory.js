import React, {useState} from 'react'
import PropTypes from "prop-types"

export const AddCategory = ( {setCategories} ) => { // aqui como se hizo q en el padre mandara algo aca lo cachamos y asi lo mandamos a llmar

    const [inputValue, setInputValue] = useState("Add a category");
    const handleInputChange = (element) => {
        setInputValue(element.target.value); //Tras hacer esto el INPUTVALUE tendra el ultimo valor escrito
    }
    const handleSubmit = (element) => {
        element.preventDefault(); /// ****** Esto es para evitar que el form provoque un refresh de la pagina
        if(inputValue.trim().length > 2){
            setCategories( categories =>  [ inputValue, ...categories] ); // obt el estado anterior y anexamos el input value
            setInputValue("");
        }
    }
    const handleClick = (element) => {
        setInputValue("");
    }
    /// en este caso no usamos el <></> para indicar que hay un elemento agrupador del componente porque el FORM ya lo es
    return (
          <form className="animate__animated animate__zoomIn animate__delay-0.5s" onSubmit={handleSubmit}>
              <p> {inputValue} </p>
            <input 
                    type="text"
                    value= {inputValue}
                    onChange={ handleInputChange }
                    onClick={handleClick}

            />
          </form>
    )
}

AddCategory.propTypes = {
        setCategories: PropTypes.func.isRequired  // es func porque el setCategories es una funcion lo que se envia
}

AddCategory.defaultProps = {
    setCategories: []
}
