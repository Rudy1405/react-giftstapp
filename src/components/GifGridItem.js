import { prototype } from 'enzyme-adapter-react-16'
import React from 'react'
import PropTypes from 'prop-types'; // shortcut con impt

export const GifGridItem = ( {id,title,url} ) => { // aqui ya tenemos los props enviados por [...img] por ende podemos llamar asi a sus elementos solo para usarlos directamente cada vez que se cree un componente de estos
   
    return (
        <div className="card animate__animated animate__fadeInUp animate__delay-0.5s">
             <img src={url? url : ""} alt={title? title : "No Title"} />
             <p>{title? title : "Imagen Generica"}</p>
        </div>
    )
}

GifGridItem.protoType = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
}