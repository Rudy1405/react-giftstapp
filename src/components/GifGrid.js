import React, {useState, useEffect} from 'react'
import PropTypes from 'prop-types';
//import { getGifs } from '../helpers/getGifs';
import { useFetchGif } from '../hooks/useFetchGif';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => {


    const {data:images, loading} = useFetchGif(category); // como usefetchGif regresa un obj, lo podemos desestructurar
          /// en la desesctructracion podemos renombrar variables, por ejemplo data llamarle images para usarlo como images abajo
/*
    El useEffect nos servia para cargar por primera vez todo, pero se uso el customHOOK de usefetchgif por ende ya no se ocupara

    const [imagesValue, setImagesValue] = useState([])
    useEffect( ()=>{ // useEffect es un hook para ejecutar codigo de forma condicional para que el getgifs por cualqueir cambio no se este llame y llame pueh
        getGifs(category) // mando a llamar a mi helper que es una promesa que me hace el fetch de lo que mande en category desde el AddCategory
        .then(imgs => setImagesValue(imgs)) 
    }, [category]) // aqui van las dependencias de cuando ejecutar el codigo... un [] es para que solo se ejecute 1 vez
                   // se pone con [category] para evitar un warning de que el category puede cambiar en alguna de sus posiciones 
     /// antes aqui estaba el getgifs pero se paso a un helper para tener mejor acomodado el codigo, ahora se llama drecto en el useEffect
*/
   



    return (
        <>
            <h3 className="animate__animated animate__fadeInUp"> {category} </h3>

            <div className="container">
                <div className="row">
                    {loading && <p className="animate__animated animate__hinge animate__delay-0.5s">Loading...</p>}    { /*  lo anterior es igual a {loading ? <p>Loading...</p> : null} */}
                    <div className='class="col-6"' className="cardGrid">            
                        { // las {} es para denotar que vamos a meter JS
                            images.map( img => { // en lugar de (img) se puede desestructurar a {id,title,url} para solo tener los elementos POR elemento que queremos
                            return <GifGridItem 
                                            key= {img.id}
                                            {...img } // mandandolo asi, estamos mandando solo las propiedades de cada IMG y por ende poder desestructurar en el GIFGRIDITEM
                                /> // Creamos un nuevo componente porque pues adentro ira un pedazo de HTML con mucho mas cuerpo
                            })  
                        }
                    </div>
                </div>
            </div>

            { // Esto es con la version del useEfecct antes del customHook
            /* <div className="container"> 
                <div className="row">
                    <div className='class="col-6"' className="cardGrid">            
                        { // las {} es para denotar que vamos a meter JS
                            imagesValue.map( img => { // en lugar de (img) se puede desestructurar a {id,title,url} para solo tener los elementos POR elemento que queremos
                            return <GifGridItem 
                                            key= {img.id}
                                            {...img } // mandandolo asi, estamos mandando solo las propiedades de cada IMG y por ende poder desestructurar en el GIFGRIDITEM
                                /> // Creamos un nuevo componente porque pues adentro ira un pedazo de HTML con mucho mas cuerpo
                            })  
                        }
                    
                    </div>
                </div>
            </div> */}
           
            
        </>
    )
}


GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}