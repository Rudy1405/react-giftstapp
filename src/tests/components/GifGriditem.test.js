import React from "react"
import { shallow } from "enzyme"
import { GifGridItem } from "../../components/GifGridItem"

describe('Pruebas en <GifGridItem>', () => {

    const title = "Un titulo";
    const url = "https://localhost/somethong.jpg";
    const wrapper = shallow( <GifGridItem title={title} url={url}/> )

    test('should show the component correctly ', () => {

        expect(wrapper).toMatchSnapshot();

    })
    

    test('should have a paragraf with the TITLE', () => {
        const p = wrapper.find("p");
        expect(p.text().trim()).toBe(title);
    })

    test('should has src an alt in img tag', () => {
        const img = wrapper.find("img");
        //img.html() cuando no hay enzyme y se trabaja como un string ya que returns el TAG en string
        expect(img.props().src).toBe(url)
        expect(img.props().alt).toBe(title) //tambien se puede img.prop("src")
    })

    test('should have animate__fadeInUp  in div', () => {
        const div = wrapper.find("div");
        expect((div.props().className).includes("animate__fadeInUp")).toBe(true)

    })
    
    
    

})
