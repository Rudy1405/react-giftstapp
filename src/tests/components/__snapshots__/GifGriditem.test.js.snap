// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifGridItem> should show the component correctly  1`] = `
<div
  className="card animate__animated animate__fadeInUp animate__delay-0.5s"
>
  <img
    alt="Un titulo"
    src="https://localhost/somethong.jpg"
  />
  <p>
    Un titulo
  </p>
</div>
`;
